extensions-core/datasketches/src/main/java/io/druid/query/aggregation/datasketches/theta/SketchBufferAggregator.java

```diff
@@ -87,12 +81,27 @@ public Object get(ByteBuffer buf, int position)
   //Note that this is not threadsafe and I don't think it needs to be
   private Union getUnion(ByteBuffer buf, int position)
   {
-    Union union = unions.get(position);
-    if (union == null) {
-      Memory mem = new MemoryRegion(nm, position, maxIntermediateSize);
-      union = (Union) SetOperation.wrap(mem);
-      unions.put(position, union);
+    Int2ObjectMap<Union> unionMap = unions.get(buf);
+    Union union = unionMap != null ? unionMap.get(position) : null;
+    if (union != null) {
+      return union;
+    }
+    return createNewUnion(buf, position, true);
+  }
+
+  private Union createNewUnion(ByteBuffer buf, int position, boolean isWrapped)
+  {
+    NativeMemory nm = getNativeMemory(buf);
+    Memory mem = new MemoryRegion(nm, position, maxIntermediateSize);
+    Union union = isWrapped
+                  ? (Union) SetOperation.wrap(mem)
+                  : (Union) SetOperation.builder().initMemory(mem).build(size, Family.UNION);
+    Int2ObjectMap<Union> unionMap = unions.get(buf);
```

Comment:
> I think enforcing the same buffer by buffer == expectedBuffer and getting rid of maps is the best solution now.

This would break BufferGrouper, which is used by groupBy, see https://github.com/druid-io/druid/pull/4071#discussion_r107326259.

I really don't see anything in the BufferAggregator javadocs that suggests the ByteBuffer passed in must always be the same. But apparently a few people (I think @leventov, @cheddar, maybe others) have read them this way. So the javadocs should at least be clarified one way or the other. And if they're clarified to outlaw providing different buffers, then BufferGrouper should change somehow, since it's bad for some Druid code to be violating contracts of other Druid code.